subsys/caf/modules/sensor_data_aggregator.c

```diff
@@ -0,0 +1,209 @@
+#include <caf/events/sensor_aggregator_event.h>
+#include <caf/sensor_manager.h>
+
+#define MODULE sensor_data_aggregator
+#include <caf/events/module_state_event.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_CAF_SENSOR_MANAGER_LOG_LEVEL);
+
+#define DT_DRV_COMPAT caf_aggregator
+
+#define __DEFINE_DATA(i, agg_id, size)							\
+	__ALIGNED(4) static uint8_t agg_ ## agg_id ## _buff_ ## i ## _data[size];
+
+#define __INITIALIZE_BUF(i, agg_id)							\
+	{(uint8_t *) &agg_ ## agg_id ## _buff_ ## i ## _data				\
+	},
+
+/* buf_len has to be equal to n*sensor_data_size*/
+#define __DEFINE_BUF_DATA(i)									\
+	UTIL_LISTIFY(DT_INST_PROP(i, buf_count), __DEFINE_DATA, i,				\
+		DT_INST_PROP(i, buf_data_length))						\
+	static struct aggregator_buffer agg_ ## i ## _bufs[] = {				\
+	UTIL_LISTIFY(DT_INST_PROP(i, buf_count), __INITIALIZE_BUF, i)				\
+	};											\
+	BUILD_ASSERT((DT_INST_PROP(i, buf_data_length) % DT_INST_PROP(i, sensor_data_size)) == 0,\
+		"Wrong sensor data or buffer size");
+
+
+#define __DEFINE_AGGREGATOR(i)								\
+	[i].sensor_descr = DT_INST_PROP(i, sensor_descr),				\
+	[i].sensor_data_size = DT_INST_PROP(i, sensor_data_size),			\
+	[i].buf_count = DT_INST_PROP(i, buf_count),					\
+	[i].buf_len = DT_INST_PROP(i, buf_data_length),					\
+	[i].agg_buffers = (struct aggregator_buffer *) &agg_ ## i ## _bufs,		\
+	[i].active_buf = (struct aggregator_buffer *) &agg_ ## i ## _bufs,
+
+#define __DEFINE_AGG							\
+	static struct aggregator aggregators[] = {			\
+	DT_INST_FOREACH_STATUS_OKAY(__DEFINE_AGGREGATOR)		\
+	}
+
+struct aggregator_buffer {
+	uint8_t *data;		/* Dynamic data. */
+	bool busy;		/* Buffer status. */
+	uint8_t sample_cnt;	/* Number of samples already saved in the buffer. */
+};
+
+struct aggregator {
+	const char *sensor_descr;			/* sensor_description of the sensor. */
+	struct aggregator_buffer *agg_buffers;	/* Buffers */
+	struct aggregator_buffer *active_buf;	/* Active buffer to which data will be placed */
+	enum sensor_state sensor_state;		/* Sensors state */
+	const uint8_t sensor_data_size;		/* Size of sensor data in bytes. */
+	const uint8_t buf_count;		/* Number of buffers */
+	const uint8_t buf_len;			/* Size of buffor data in bytes. */
+};
+
+
+DT_INST_FOREACH_STATUS_OKAY(__DEFINE_BUF_DATA) //no semicolon on purpose.
+__DEFINE_AGG;
+
+
+static struct aggregator_buffer *get_free_buffer(struct aggregator *agg)
+{
+	for (size_t i = 0; i < agg->buf_count; i++) {
+		if (agg->agg_buffers[i].busy == false) {
+			return &agg->agg_buffers[i];
+		}
+	}
+	return NULL;
+}
+
+static struct aggregator *get_aggregator(const char *sensor_descr)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(aggregators); i++) {
+		if (sensor_descr == aggregators[i].sensor_descr) {
+			return &aggregators[i];
+		}
+	}
+	return NULL;
+}
+
+static void clean_agg_buffer(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	__ASSERT_NO_MSG(ab);
+
+	LOG_INF("Cleaning, sample_cnt %d", ab->sample_cnt);
```

Comment:
we don't care about sample count here but we care which buffer was released
`"Releasing buffer %p", ab`

Reply:
If we want to keep this log I recommend to use DBG level for it and create a matching "Allocated buffer %p" in the get_free_buffer